### DISCUSSION TITLE
Using dependsOn causes error

### DISCUSSION THREAD

**Discussion by sergeykad** [jsonschema2dataclass/js2d-gradle#451] [Q&A] [answered]
I want to migrate from the official plugin and apart from replacing `source` with `source.setFrom` in configuration, everything went fine. That is until I replaced task dependency with the new task.  I need to execute another task before `generateJsonSchema2DataClass` but all my attempts resulted in an error.

I tried the following approaches:

1. `generateJsonSchema2DataClass.dependsOn(myTask)`
2. `tasks.generateJsonSchema2DataClass.dependsOn(myTask)`
3. `project.tasks.generateJsonSchema2DataClass.dependsOn(myTask)`
4. `generateJsonSchema2DataClass.configure { dependsOn myTask }`
5. `tasks.withType(generateJsonSchema2DataClass).configureEach {  dependsOn(myTask) }`

I am always getting the following error: `Could not get unknown property 'generateJsonSchema2DataClass' for task set of type org.gradle.api.internal.tasks.DefaultTaskContainer.` 
except for the last two options which return: 
`Could not get unknown property 'generateJsonSchema2DataClass' for project ':objects' of type org.gradle.api.Project.`

I am using Gradle v7.5

The issue can be related to https://github.com/jsonschema2dataclass/js2d-gradle/issues/13

**Comment by eirnym** [comment]
Could you explain your case a bit more than ending results? What is the goal you want to a achieve? 

The only reason I can suppose is you want to process resources (copy, download, unzip, etc) before passing them to the generator, is my guess correct or you have a bit more complicated scenario I haven't thought about?

PS: if it's a simple resource processing, plugin is dependent on the `ProcessingResources` task

**Comment by sergeykad** [comment]
We have a custom task that preprocesses JSON files before they are fed to the `JsonSchema2pojo`. 

I checked it now and our task is indeed executed since `ProcessingResources` depends on it. I see that there is no issue after all and I just have to remove the `generateJsonSchema2Pojo.dependsOn(myTask)` string instead of changing it.

Thanks a lot.

**Comment by eirnym** [comment]
Thank you too!

**Comment by sergeykad** [comment]
I was too quick to close the issue. I missed it previously, but there is the following warning from Gradle:
>    Gradle detected a problem with the following location: '/home/user/project/build/generated/sources/js2d/main'. Reason: Task ':project:sourcesJar' uses this output of task ':project:generateJsonSchema2DataClass0' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5/userguide/validation_problems.html#implicit_dependency for more details about this problem.

Adding `sourcesJar.dependsOn(generateJsonSchema2DataClass)` results in the same error as mentioned previously.
The same dependency works fine with the original plugin. In any case, it is strange that the plugin does not support adding dependencies.

**Comment by eirnym** [comment, marked as answer]
Correct, you can't hook on the generator task itself. Currently, all tasks are created on `afterEvaluate` stage. This could be changed later on if more precision is required. 

Generator task depends on `ProcessingResources` and lombok and compile tasks (java and kotlin) depend on generator task. 

Thus you may hook either before `ProcessingResources` or after `build`. 

To pack schemas and sources I use following scripts for a project where many models built at once:

```gradle
// part of jsongen.groovy
task copySchemas(type: Copy) {
   from jsonDirBase            // original schema placement
   into targetJSONBaseDir // some random directory under build folder
}
project.tasks.build.dependsOn(copySchemas) // it's safe to copy independently
```


```gradle
// part of publishing.groovy

task sourceJar(type: Jar, dependsOn: build) {
   duplicatesStrategy = DuplicatesStrategy.EXCLUDE
  from sourceSets.main.allJava
  archieveClassifier.set "sources"
}

task schemasJar(type: Jar, dependsOn: build) {
   duplicatesStrategy = DuplicatesStrategy.EXCLUDE
  from targetJSONBaseDir
  archieveClassifier.set "schemas"
}

publishing {
  publications {
     modulePublish(MavenPublications) {
        artifact jar
        artifact sourcesJar
        artifact schemasJar
        artifactId = project.artifactId
     }
  }
}
```

**Comment by eirnym** [comment]
Did this a trick for you?

**Comment by sergeykad** [comment]
I meanwhile gave up on creating source jars. I will try your solution if I'll see that I have to create them.
Thanks in any case. The support is appreciated.

**Comment by eirnym** [comment]
I've converted to a discussion, as it's not an issue

**Comment by eirnym** [comment]
Any update on the topic?

**Comment by eirnym** [comment]
@sergeykad For java, groovy or kotlin (not-android) projects you can actually use `dependsOn` from now on. I'll investigate how I can push forward it for android projects as well

UPD: this is enabled for all tasks, released in 5.0

**Comment by eirnym** [comment]
@sergeykad Added [an example](https://github.com/jsonschema2dataclass/js2d-gradle/tree/main/demo/java/model-publish) how to efficinlty pack sources and schemas to jars and publish them.
